src/lib/uptime_service/uptime_service.ml

```diff
@@ -0,0 +1,455 @@
+                        in
+                        ( match
+                            Network_pool.Snark_pool.Resource_pool.add_snark
+                              snark_resource_pool ~is_local:true ~work
+                              ~proof:(`One proof) ~fee
+                          with
+                        | `Added ->
+                            [%log info]
+                              "Added work created for uptime service to SNARK \
+                               pool"
+                        | `Statement_not_referenced ->
+                            [%log error]
+                              "Statement not referenced error when trying to \
+                               add SNARK for uptime service to pool" ) ;
+                        let proof_data = {proof; proof_time; snark_work_fee} in
+                        let proof_len =
+                          bin_write_proof_data proof_buf ~pos:0 proof_data
+                        in
+                        let proof_string =
+                          String.init proof_len ~f:(fun ndx -> proof_buf.{ndx})
+                        in
+                        match Base64.encode proof_string with
+                        | Error (`Msg err) ->
+                            [%log error]
+                              "Could not Base64-encode SNARK work for uptime \
+                               service"
+                              ~metadata:[("error", `String err)] ;
+                            return ()
+                        | Ok snark_work_base64 ->
+                            let state_hash =
+                              Transition_frontier.Breadcrumb.state_hash
+                                best_tip
+                            in
+                            let block_data =
+                              { block= block_base64
+                              ; created_at= get_rfc3339_time ()
+                              ; peer_id
+                              ; snark_work= Some snark_work_base64 }
+                            in
+                            send_uptime_data ~logger ~submitter_keypair ~url
+                              ~state_hash block_data ) ) ) ) )
+
+let start (mina : Mina_lib.t) =
+  let config = Mina_lib.config mina in
+  let transition_frontier = Mina_lib.transition_frontier mina in
+  match config.uptime_url with
+  | None ->
+      [%log' info config.logger] "Not running uptime service, no URL given" ;
+      ()
+  | Some url ->
+      [%log' info config.logger] "Starting uptime service using URL $url"
+        ~metadata:[("url", `String (Uri.to_string url))] ;
+      let slot_duration_ms =
+        Consensus.Configuration.t
+          ~constraint_constants:Genesis_constants.Constraint_constants.compiled
+          ~protocol_constants:Genesis_constants.compiled.protocol
+        |> Consensus.Configuration.slot_duration |> Float.of_int
+      in
+      let five_slots_span = Time.Span.of_ms (slot_duration_ms *. 5.0) in
```

Comment:
Does it make sense to use `4.5` here, to try to counteract drift? I think `Clock.every'` waits for the function to complete before scheduling the next one.

Perhaps we should also use a `don't_wait_for` and an `Interruptible.t` below, so that we know we're not delaying scheduling of the check, but can still guarantee we don't have 2 submission attempts in flight at once?